pkg/repository/network-validation.go

```diff
@@ -0,0 +1,536 @@
+		return fmt.Errorf("private super network must always contain a defaultchildprefixlength")
+	}
+
+	for _, af := range prefixes.AddressFamilies() {
+		if _, exists := defaultChildPrefixLength[af]; !exists {
+			return fmt.Errorf("private super network must always contain a defaultchildprefixlength per addressfamily:%s", af)
+		}
+	}
+
+	return nil
+}
+
+func (r *networkRepository) validateAdditionalAnnouncableCIDRs(additionalCidrs []string, privateSuper bool) error {
+	if len(additionalCidrs) == 0 {
+		return nil
+	}
+
+	if !privateSuper {
+		return errors.New("additionalannouncablecidrs can only be set in a private super network")
+	}
+
+	for _, cidr := range additionalCidrs {
+		_, err := netip.ParsePrefix(cidr)
+		if err != nil {
+			return fmt.Errorf("given cidr:%q in additionalannouncablecidrs is malformed:%w", cidr, err)
+		}
+	}
+
+	return nil
+}
+func (r *networkRepository) ValidateUpdate(ctx context.Context, req *adminv2.NetworkServiceUpdateRequest) (*Validated[*adminv2.NetworkServiceUpdateRequest], error) {
+	old, err := r.Get(ctx, req.Id)
+	if err != nil {
+		return nil, err
+	}
+	newNetwork := *old
+
+	if req.DefaultChildPrefixLength != nil && !old.PrivateSuper {
+		return nil, errorutil.InvalidArgument("defaultchildprefixlength can only be set on privatesuper")
+	}
+
+	if old.ParentNetworkID != "" || (old.NetworkType != nil && *old.NetworkType == metal.ChildNetworkType) {
+		if len(req.Prefixes) > 0 {
+			return nil, errorutil.InvalidArgument("cannot change prefixes in child networks")
+		}
+	}
+
+	var (
+		prefixesToBeRemoved metal.Prefixes
+		prefixesToBeAdded   metal.Prefixes
+		destPrefixAfs       metal.AddressFamilies
+	)
+
+	prefixesToBeRemoved, prefixesToBeAdded, err = r.calculatePrefixDifferences(ctx, old, &newNetwork, req.Prefixes)
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	r.r.log.Debug("validate update", "old parent", old.ParentNetworkID, "prefixes to remove", prefixesToBeRemoved, "prefixes to add", prefixesToBeAdded)
+	// Do not allow to change prefixes on child networks
+	if old.ParentNetworkID != "" && (len(prefixesToBeRemoved) > 0 || len(prefixesToBeAdded) > 0) {
+		return nil, errorutil.InvalidArgument("cannot change prefixes in child networks")
+	}
+
+	if req.DestinationPrefixes != nil {
+		destPrefixes, err := metal.NewPrefixesFromCIDRs(req.DestinationPrefixes)
+		if err != nil {
+			return nil, errorutil.Convert(err)
+		}
+		destPrefixAfs = destPrefixes.AddressFamilies()
+	}
+
+	err = r.validatePrefixesAndAddressFamilies(newNetwork.Prefixes, destPrefixAfs, old.DefaultChildPrefixLength, old.PrivateSuper)
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	_, err = metal.ToChildPrefixLength(req.DefaultChildPrefixLength, newNetwork.Prefixes)
+	if err != nil {
+		return nil, errorutil.NewInvalidArgument(err)
+	}
+	_, err = metal.ToChildPrefixLength(req.MinChildPrefixLength, newNetwork.Prefixes)
+	if err != nil {
+		return nil, errorutil.NewInvalidArgument(err)
+	}
```

Comment:
Not sure if I have missed it but doesn't this allow child networks to sneak in fields? With the nice validations that were added I think all these have to be called again for the update to ensure all of it. Like do another switch on the network type and call the validations for this type.